src/loaders/workspace_uploader/workspace_uploader.py

```diff
@@ -443,6 +457,56 @@ def _upload_assembly_files_in_parallel(
     return failed_names
 
 
+def _gen(
+        conf: Conf,
+        wait_to_upload_assemblies: dict[str, str]
+) -> Generator[Tuple[tuple[str], tuple[str], tuple[str], int], None, None]:
+    """
+    Generator function to yield the assembly files to upload.
+    """
+    assembly_files_len = len(wait_to_upload_assemblies)
+    assembly_names = tuple(wait_to_upload_assemblies.keys())
+    host_assembly_dirs = tuple(wait_to_upload_assemblies.values())
+    container_internal_assembly_paths = tuple(
+        [
+            os.path.join(JOB_DIR_IN_ASSEMBLYUTIL_CONTAINER, assembly_name)
+            for assembly_name in assembly_names
+        ]
+    )
+
+    start_idx = 0
+    cumsize = 0
+    # to avoid memory issue, upload 1GB * MEMORY_UTILIZATION of assembly files at a time
+    max_cumsize = 1000000000 * MEMORY_UTILIZATION
+    for idx in range(assembly_files_len):
+        file_path = os.path.join(conf.job_data_dir, assembly_names[idx])
+        if not os.path.exists(file_path):
+            raise ValueError(f"{file_path} does not exist. "
+                             f"Ensure file {assembly_names[idx]} exist in {conf.job_data_dir}")
+        file_size = os.path.getsize(file_path)
+        # cumulate aasembly files until the total size is greater than max_cumsize
+        if file_size + cumsize <= max_cumsize:
+            cumsize += file_size
+        else:
+            yield (
+                container_internal_assembly_paths[start_idx:idx],
+                host_assembly_dirs[start_idx:idx],
+                assembly_names[start_idx:idx],
+                idx - start_idx,
+            )
+            # reset start_idx and cumsize
+            start_idx = idx
+            cumsize = file_size
+
+    # yield the remaining assembly files
+    yield (
```

Comment:
we don't need to explicitly check `if start_idx < assembly_files_len` before yielding the remaining assembly files?

Reply:
My understanding is start_idx is in range(assembly_files_len). So no explicitly check is needed.

Reply:
👍